lib/rubocop/magic_comment.rb

```diff
@@ -0,0 +1,186 @@
+    TOKEN = /[[:alnum:]\-_]+/
+
+    # Detect magic comment format and pass it to the appropriate wrapper.
+    #
+    # @param comment [String]
+    #
+    # @return [RuboCop::MagicComment]
+    def self.parse(comment)
+      case comment
+      when Emacs::FORMAT then Emacs.new(comment)
+      when Vim::FORMAT   then Vim.new(comment)
+      else
+        Simple.new(comment)
+      end
+    end
+
+    def initialize(comment)
+      @comment = comment
+    end
+
+    # Does the magic comment enable the frozen string literal feature.
+    #
+    # Test whether the frozen string literal value is `true`. Cannot
+    # just return `frozen_string_literal` since an invalid magic comment
+    # like `# frozen_string_literal: yes` is possible and the truthy value
+    # `'yes'` does not actually enable the feature
+    #
+    # @return [Boolean]
+    def frozen_string_literal?
+      frozen_string_literal == true
+    end
+
+    # Was a magic comment for the frozen string literal found?
+    #
+    # @return [Boolean]
+    def frozen_string_literal_specified?
+      !frozen_string_literal.nil?
+    end
+
+    def frozen_string_literal
+      return unless (setting = extract_frozen_string_literal)
+
+      case setting
+      when 'true'  then true
+      when 'false' then false
+      else
+        setting
+      end
+    end
+
+    private
+
+    # Match the entire comment string with a pattern and take the first capture.
+    #
+    # @param pattern [Regexp]
+    #
+    # @return [String] if pattern matched
+    # @return [nil] otherwise
+    def extract(pattern)
+      @comment[pattern, 1]
+    end
+
+    # Parent to Vim and Emacs magic comment handling.
+    #
+    # @abstract
+    class EditorComment < MagicComment
+      private
+
+      # Find a token starting with the provided keyword and extract its value.
+      #
+      # @param keyword [String]
+      #
+      # @return [String] extracted value if it is found
+      # @return [nil] otherwise
+      def match(keyword)
+        pattern = /\A#{keyword}\s*(#{TOKEN})\z/
+
+        tokens.each do |token|
+          next unless (value = token[pattern, 1])
+
+          return value.downcase
+        end
+
+        nil
+      end
+
+      # Individual tokens composing an editor specific comment string.
+      #
+      # @return [Array<String>]
+      def tokens
+        extract(self.class::FORMAT).split(self.class::SEPARATOR).map(&:strip)
+      end
+    end
+
+    # Wrapper for emacs style magic comments.
+    #
+    # @example emacs style comment
+    #   comment = RuboCop::MagicComment.parse(
+    #     '# -*- encoding: ASCII-8BIT; frozen_string_literal: true -*-'
+    #   )
+    #
+    #   comment.encoding              # => 'ascii-8bit'
+    #   comment.frozen_string_literal # => true
+    #
+    # @see https://www.gnu.org/software/emacs/manual/html_node/emacs/Specify-Coding.html
+    # @see https://git.io/vMCXh Emacs handling in Ruby's parse.y
+    class Emacs < EditorComment
```

Comment:
Updated. I usually don't like repeating a word a bunch of times in a namespace (`MagicComment::EmacsComment` vs `MagicComment::Emacs`) but it is more consistent here.